### PR TITLE
fix: unexpected end of input error when running fluttium test

### LINKED ISSUE #29: fix: Unexpected end of input error when running fluttium test
**Description**

An error is thrown when trying to run `fluttium test`, see the stack trace below. It seems to be an error during the intial pre test "Installing dependencies" phase.

Downloaded using `flutter pub global activate fluttium_cli` current version is `0.1.0-dev.5`

Flutter version is `3.3.4` and the target device is iOS.

```
giles@macbook mobile % fluttium test my_flow.yaml
✓ Retrieving devices (4.9s)
⠦ Installing dependencies... (1.3s)Unexpected end of input
#0      _ChunkedJsonParser.fail (dart:convert-patch/convert_patch.dart:1383:5)
#1      _ChunkedJsonParser.close (dart:convert-patch/convert_patch.dart:501:7)
#2      _parseJson (dart:convert-patch/convert_patch.dart:36:10)
#3      JsonDecoder.convert (dart:convert/json.dart:612:36)
#4      JsonCodec.decode (dart:convert/json.dart:216:41)
#5      jsonDecode (dart:convert/json.dart:155:10)
#6      FluttiumRunner._setupProject (package:fluttium_runner/src/fluttium_runner.dart:134:25)
<asynchronous suspension>
#7      FluttiumRunner.run (package:fluttium_runner/src/fluttium_runner.dart:202:5)
<asynchronous suspension>
#8      TestCommand.run (package:fluttium_cli/src/commands/test_command.dart:246:5)
<asynchronous suspension>
#9      CommandRunner.runCommand (package:args/command_runner.dart:209:13)
<asynchronous suspension>
#10     FluttiumCommandRunner.runCommand (package:fluttium_cli/src/command_runner.dart:108:18)
<asynchronous suspension>
#11     FluttiumCommandRunner.run (package:fluttium_cli/src/command_runner.dart:60:14)
<asynchronous suspension>
#12     main (file:///Users/giles/.pub-cache/hosted/pub.dartlang.org/fluttium_cli-0.1.0-dev.5/bin/fluttium.dart:6:24)
<asynchronous suspension>


Fluttium, a user flow testing tool for Flutter.

Usage: fluttium <command> [arguments]

Global options:
-h, --help            Print this usage information.
-v, --version         Print the current version.
    --[no-]verbose    Noisy logging, including all shell commands executed.

Available commands:
  create   Create a FluttiumFlow test.
  test     Run a FluttiumFlow test.
  update   Update the CLI.

Run "fluttium help <command>" for more information about a command.
```

**Steps To Reproduce**

1. Create the following `my_flow.yaml` file
```yaml
description: Login
---
- tapOn: "email"
```
3. Run fluttium test my_flow.yaml
5. See error

**Expected Behavior**

The test runs as expected.

**Additional context**

Flutter doctor --verbose

```
[✓] Flutter (Channel stable, 3.3.4, on macOS 12.5.1 21G83 darwin-x64, locale en-GB)
    • Flutter version 3.3.4 on channel stable at /Users/giles/fvm/versions/2.10.3
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision eb6d86ee27 (9 days ago), 2022-10-04 22:31:45 -0700
    • Engine revision c08d7d5efc
    • Dart version 2.18.2
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/giles/Library/Android/sdk
    • Platform android-31, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.72.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (3 available)
    • iPhone 14 Pro (mobile) • 6A00D6B8-A02C-4FAB-B8BA-E5021707C1B0 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-0 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64     • macOS 12.5.1 21G83 darwin-x64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 106.0.5249.119

[✓] HTTP Host Availability
    • All required HTTP hosts are available
```

### PR DESCRIPTION
**Pull request by wolfenrain**
## Status

**READY**

## Description

Fixes #29 and makes getting JSON's more safe overal.

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [x] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

**Review by felangel — APPROVED** [review]

**Comment by wolfenrain** [comment]
@gilescm @bartekpacia this should solve the problem Fluttium had with parsing JSON in malformed strings, would love to have some eyes on this (but do not feel obligated).

Decoding it like this will allow the tool to do it for any Flutter command that technically could add extra info surrounding the JSON that we dont want.

**Review by bartekpacia — DISMISSED** [review]
I think it's great.

File: packages/fluttium_flow/lib/src/json_decode_safely.dart [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  for (var i = 0; i < source.length; i++) {
+    final currentChar = source[i];
+    final previousChar = i > 0 ? source[i - 1] : null;
+
+    buffer.write(currentChar);
+
+    // Match the opening and closing brackets, if the current character is a
+    // the open bracket we increment the depth, if it is the closing bracket
+    // we decrement the depth.
+    if (currentChar == openBracket && previousChar != r'\') {
```
Comment (bartekpacia) [inline packages/fluttium_flow/lib/src/json_decode_safely.dart:42]: Why are we checking if the 'previousChar` is not a backslash?

Comment (gilescm) [reply packages/fluttium_flow/lib/src/json_decode_safely.dart:42]: I'm guessing but could it be to ensure escaped brackets e.g. "\{" which could be present and part of a string value don't inadvertently increment the depth?

File: packages/fluttium_flow/test/src/json_decode_safely_test.dart [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+]
+
+Some output
+''';
+      final output = jsonDecodeSafely(input);
+      expect(
+        output,
+        equals([1, 2, 3]),
+      );
+    });
```
Comment (gilescm) [inline packages/fluttium_flow/test/src/json_decode_safely_test.dart:42]: Not necessary at all but a good test on the depth aspect could be:

```dart
    test('safely decodes nested JSON from a malformed string', () {
      const input = '''
{
  "a": [
    1,
    2,
    3
  ],
  "b": {
    "d": 5,
    "e": 6,
    "f": [7. 8, 9]
  },
  "c": 4
}
Some output
''';
      final output = jsonDecodeSafely(input);
      expect(
        output,
        equals({
          'a': [1, 2, 3],
          'b': {
            'd': 5,
            'e': 6,
            'f': [7, 8, 9]
          },
          'c': 4,
        }),
      );
    });
```

Comment (wolfenrain) [reply packages/fluttium_flow/test/src/json_decode_safely_test.dart:42]: Good call! Added the test! Thank you so much

**Review by bartekpacia — APPROVED** [review]

**Pull request merged**